File: OrchidCore/src/main/java/com/eden/orchid/impl/themes/functions/LinkFunction.java
package com.eden.orchid.impl.themes.functions;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.compilers.TemplateFunction;
import com.eden.orchid.api.options.annotations.Description;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.theme.pages.OrchidPage;
import lombok.Getter;
import lombok.Setter;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
@Getter @Setter
public final class LinkFunction extends TemplateFunction {

    private final OrchidContext context;

    @Option
    @Description("The Id of an item to link to.")
    private String itemId;

    @Option
    @Description("The type of collection the item is expected to come from.")
    private String collectionType;

    @Option
    @Description("The specific Id of the given collection type where the item is expected to come from.")
    private String collectionId;

    @Inject
    public LinkFunction(OrchidContext context) {
        super("link", true);
        this.context = context;
    }

    @Override
    public String[] parameters() {
        return new String[] {"itemId", "collectionId", "collectionType"};
    }

    @Override
    public Object apply() {
        OrchidPage page = context.findPage(collectionType, collectionId, itemId);
        if(page != null) {
            return page.getLink();
        }

        return "";
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/themes/functions/LinkFunction.kt
package com.eden.orchid.impl.themes.functions

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.compilers.TemplateFunction
import com.eden.orchid.api.indexing.IndexService
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.theme.pages.OrchidPage

@Description(value = "Get the link to a page.", name = "Link")
class LinkFunction : TemplateFunction("link", true) {

    @Option
    @Description("The Id of an item to link to.")
    lateinit var itemId: String

    @Option
    @Description("The type of collection the item is expected to come from.")
    lateinit var collectionType: String

    @Option
    @Description("The specific Id of the given collection type where the item is expected to come from.")
    lateinit var collectionId: String

    override fun parameters(): Array<String> {
        return IndexService.locateParams
    }

    override fun apply(context: OrchidContext, page: OrchidPage?): Any? {
        val foundPage = context.findPage(collectionType, collectionId, itemId)
        return if (foundPage != null) {
            foundPage.link
        } else ""

    }
}
